File: apps/codemap-search/tests/fixtures/navigation/typescript/policy.ts
import type { Rule } from "./rules";

type RuleResult<T> = T extends { name: infer Name } ? Name : never;

export class OrderPolicy<T extends Rule> {
  constructor(private readonly rules: readonly T[]) {}

  evaluate(order: Order): RuleResult<T>[] {
    const active = this.rules.filter((rule) => rule.check(order));
    const names = active.map((rule) => {
      const { name: ruleName, severity = "low", ...metadata } = rule as T & {
        severity?: string;
      };
      return `${ruleName}:${severity}:${Object.keys(metadata).length}`;
    });
    return names as RuleResult<T>[];
  }
}

File: apps/codemap-search/tests/fixtures/navigation/python/policy.py
from typing import Protocol


class Rule(Protocol):
    def check(self, order) -> bool: ...


def evaluate(order, rules: list[Rule]) -> list[str]:
    failures: list[str] = []
    for rule in rules:
        passed = rule.check(order)
        if not passed:
            failures.append(rule.__class__.__name__)

    all(rule.check(order) for rule in rules)
    return failures
